sdk/core/azure_core/src/http/policies/bearer_token_policy.rs

```diff
@@ -59,24 +70,143 @@ impl Policy for BearerTokenAuthorizationPolicy {
         request: &mut Request,
         next: &[Arc<dyn Policy>],
     ) -> PolicyResult {
-        let access_token = self.access_token.read().await;
+        self.on_request
+            .on_request(ctx, request, self.authorizer.as_ref())
+            .await?;
 
+        let mut response = next[0].send(ctx, request, &next[1..]).await?;
+
+        if response.status() == StatusCode::Unauthorized {
+            self.authorizer.invalidate_cache().await;
+            if let Some(ref on_challenge) = self.on_challenge {
+                let has_challenge = response
+                    .headers()
+                    .get_optional_str(&WWW_AUTHENTICATE)
+                    .is_some();
+                if has_challenge {
+                    let should_retry = on_challenge
+                        .on_challenge(ctx, request, self.authorizer.as_ref(), response.headers())
+                        .await?;
+                    if should_retry {
+                        #[cfg(not(target_arch = "wasm32"))]
+                        if let SeekableStream(stream) = request.body_mut() {
+                            stream.reset().await?;
+                        }
+                        response = next[0].send(ctx, request, &next[1..]).await?
+                    }
+                }
+            }
+        }
+
+        Ok(response)
+    }
+}
+
+/// Callback [`BearerTokenAuthorizationPolicy`] invokes when it receives a 401 Unauthorized response with an authentication challenge (WWW-Authenticate header).
+#[cfg_attr(target_arch = "wasm32", async_trait(?Send))]
+#[cfg_attr(not(target_arch = "wasm32"), async_trait)]
+pub trait OnChallenge: std::fmt::Debug + Send + Sync {
+    /// Called when [`BearerTokenAuthorizationPolicy`] receives a 401 Unauthorized response with a challenge. Implementations are
+    /// responsible for parsing authentication parameters from the challenge, authorizing the request via the provided [`Authorizer`],
+    /// and indicating whether the policy should retry the request.
+    ///
+    /// # Arguments
+    /// * `context` - The request context
+    /// * `request` - The HTTP request that received the challenge
+    /// * `authorizer` - Helper used to acquire an access token and set the request's authorization header
+    /// * `headers` - The 401 response's headers
+    ///
+    /// # Returns
+    /// * `Ok(true)` when the callback handled the challenge and [`BearerTokenAuthorizationPolicy`] should retry the request.
+    /// * `Ok(false)` when the callback can't handle the challenge. [`BearerTokenAuthorizationPolicy`] will return the 401 response to the client in this case.
+    /// * `Err` when an error occurs while handling the challenge.
+    async fn on_challenge(
+        &self,
+        context: &Context,
+        request: &mut Request,
+        authorizer: &dyn Authorizer,
+        headers: &Headers,
+    ) -> Result<bool>;
```

Comment:
And a `bool` should always be sufficient? You can't imagine expanding on that such that maybe we should return a result model that is currently just a bool? If you're reasonably certain, I see no problem with this.